include/zephyr/device.h

```diff
@@ -943,6 +957,9 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
 	Z_DEVICE_INIT_ENTRY_DEFINE(dev_id, init_fn, level, prio)
 
 #if defined(CONFIG_HAS_DTS) || defined(__DOXYGEN__)
+/* TODO: remove this, applications should explicitely declare so they can
+ * obtain friendly compile time errors
```

Comment:
You don't actually want to do this IMO. We tried not having it at first and it really is required. Removing this would require explicit declarations inside of the macros used for device instantiation in device drivers. This is extra boilerplate that I think we don't want.

Reply:
If this approach moves forward, this is the only way to actually enforce the build asserts. Of course, we can have a Kconfig option to enable this hidden magic (maybe a good way to deprecate or keep it as a facility for those who don't care about good diagnostics).